homeassistant/components/mazda/binary_sensor.py

```diff
@@ -0,0 +1,184 @@
+"""Platform for Mazda binary sensor integration."""
+from __future__ import annotations
+
+from collections.abc import Callable
+from dataclasses import dataclass
+
+from homeassistant.components.binary_sensor import (
+    BinarySensorDeviceClass,
+    BinarySensorEntity,
+    BinarySensorEntityDescription,
+)
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.core import HomeAssistant
+from homeassistant.helpers.entity_platform import AddEntitiesCallback
+
+from . import MazdaEntity
+from .const import DATA_CLIENT, DATA_COORDINATOR, DOMAIN
+
+
+@dataclass
+class MazdaBinarySensorRequiredKeysMixin:
+    """Mixin for required keys."""
+
+    # Suffix to be appended to the vehicle name to obtain the binary sensor name
+    name_suffix: str
+
+    # Function to determine the value for this binary sensor, given the coordinator data
+    value: Callable[[dict], bool]
+
+
+@dataclass
+class MazdaBinarySensorEntityDescription(
+    BinarySensorEntityDescription, MazdaBinarySensorRequiredKeysMixin
+):
+    """Describes a Mazda binary sensor entity."""
+
+    # Function to determine whether the vehicle supports this binary sensor, given the coordinator data
+    is_supported: Callable[[dict], bool] = lambda data: True
+
+    # Function to return extra state attributes, given the coordinator data
+    state_attributes: Callable[[dict], dict] | None = None
+
+
+def _plugged_in_supported(data):
+    """Determine if 'plugged in' binary sensor is supported."""
+    return (
+        data["isElectric"] and data["evStatus"]["chargeInfo"]["pluggedIn"] is not None
+    )
+
+
+def _charging_supported(data):
+    """Determine if 'charging' binary sensor is supported."""
+    return data["isElectric"] and data["evStatus"]["chargeInfo"]["charging"] is not None
+
+
+def _doors_open(data):
+    """Determine if at least one vehicle door is open."""
+    door_status = data["status"]["doors"]
+    return (
+        door_status["driverDoorOpen"]
+        or door_status["passengerDoorOpen"]
+        or door_status["rearLeftDoorOpen"]
+        or door_status["rearRightDoorOpen"]
+    )
+
+
+def _trunk_open(data):
+    """Determine if the vehicle trunk is open."""
+    return data["status"]["doors"]["trunkOpen"]
+
+
+def _hood_open(data):
+    """Determine if the vehicle hood is open."""
+    return data["status"]["doors"]["hoodOpen"]
+
+
+def _vehicle_is_plugged_in(data):
+    """Determine if vehicle is plugged in."""
+    return data["evStatus"]["chargeInfo"]["pluggedIn"]
+
+
+def _vehicle_is_charging(data):
+    """Determine if vehicle is charging."""
+    return data["evStatus"]["chargeInfo"]["charging"]
```

Comment:
This isn't reused? Why not just put that directly into the lambda in the entity description?

(Applies to a bunch of more of these).

Reply:
I've converted functions to lambdas in cases where the lambda would be a single line. I was asked in https://github.com/home-assistant/core/pull/63423 to avoid multi-line lambdas, so I ignored the ones that would be multi-line lambdas.